### PR TITLE
[DOC] New file RI.md

### PR DESCRIPTION
**Pull request by BurdetteLamar**
Replaces the very spare `RI.rdoc` with the robust `RI.md`.

This is meant for new or less-experienced RI users (who may have found this page in a browser search).

### CODE REVIEW

**Review by peterzhu2118 — COMMENTED** [review]
I find this document to be very wordy and repetitive. I have a few concerns:

1. Who is this targeted towards? Users of `ri` want a quick reference and are probably using the `ri -h` or the `man ri` documentation.
2. There are many repeated parts in this documentation. The "Class and Module Documents" section on line 326 is pretty similar to the "Ruby Class and Module Documents" section on 572. The documentation for static and interactive are very similar and are repeats of one another.

**Comment by BurdetteLamar** [comment]
>     1. Who is this targeted towards? Users of `ri` want a quick reference and are probably using the `ri -h` or the `man ri` documentation.

I'm thinking more of the new user (possibly the never-before-user).  This content will need to be someplace where both the user and the search engines will find it.

An eventual aim is to show (in -h and --help) ri commands that will fetch relevant documentation.

I'd never heard of `man ri`.  It's very old (and no doubt out-of-date).  Seems to be actual source (not derived);  is that right?  I find that it's available on my Linux machine, but not on Windows;  don't know about elsewhere.

**Comment by BurdetteLamar** [comment]
>     2. There are many repeated parts in this documentation. The "Class and Module Documents" section on line 326 is pretty similar to the "Ruby Class and Module Documents" section on 572. The documentation for static and interactive are very similar and are repeats of one another.

Some of the repetition arises from my misunderstanding about how many things are here.  I'd thought four (class/page ruby/gem), but it may be more useful to think of two (class/page only).  All classes (including gem classes) are already available in the class list; if we think of `ruby:` is a pseudo-gem, all pages are of one kind.

I'll look into how to avoid repetition between static/interactive.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Done.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Fixed.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Fixed.

File: RI.md
Comment (BurdetteLamar) [reply RI.md:247]: My aim here is to show a program name with options.  So, stet?

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Note removed.

**Comment by peterzhu2118** [comment]
> I'd never heard of man ri. It's very old (and no doubt out-of-date). Seems to be actual source (not derived); is that right? I find that it's available on my Linux machine, but not on Windows; don't know about elsewhere.

I'm not sure what you mean by "actual source (not derived)". The source is [in the file /man/ri.1](https://github.com/ruby/rdoc/blob/master/man/ri.1). I think it's better to improve the man page rather than this document since man pages are the de facto way of reading the documentation for commands in Linux.

I'm not sure that it's out-of-date because the ri command hasn't really changed in the past few years.

**Comment by BurdetteLamar** [comment]
> I'm not sure what you mean by "actual source (not derived)". The source is [in the file /man/ri.1](https://github.com/ruby/rdoc/blob/master/man/ri.1).

Just meant source text -- not derived from something else; e.g., .{html|ri} derived from .{c|rb|rdoc|md}.

> I think it's better to improve the man page rather than this document since man pages are the de facto way of reading the documentation for commands in Linux.

I'm willing to work with the man page later on.  This page RI.md (though it still needs major work) is for everyone, not just for those on Linux.

**Comment by BurdetteLamar** [comment]
Much more work to be done here, so marking as Draft and removing review requests.

File: RI.md
Comment (nobu) [reply RI.md]: What about:

```sh
RI_PAGER="grep '^== '" ri Array
```

**Comment by BurdetteLamar** [comment]
@peterzhu2118, if this PR does not interest you, I can try to find another primary reviewer.

**Comment by peterzhu2118** [comment]
Sorry, I've been very busy recently with various things at work and this PR is very large, so I haven't been able to spend enough time on this. I will try to look at this when I have time, but I can't guarantee when that will be.

**Comment by BurdetteLamar** [comment]
> Sorry, I've been very busy recently with various things at work and this PR is very large, so I haven't been able to spend enough time on this. I will try to look at this when I have time, but I can't guarantee when that will be.

All good,  Will be away through April 20 (celebrating 25th wedding anniversary).

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Fixed.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Fixed.

File: RI.md
Comment (BurdetteLamar) [reply RI.md:137]: Inline note would be (IMO) too big for a table cell;  I have just omitted it entirely.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Removed.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Removed.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Removed.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Removed.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Removed.

File: RI.md
Comment (BurdetteLamar) [reply RI.md:708]: I'm reluctant to remove just some of these (the ones that don't contribute).  Leaving all of them in has the virtue of their showing in the TOC (for quick user nav).

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Fixed.

File: RI.md
Comment (peterzhu2118) [inline RI.md:134]:
```suggestion
| 'IO#readlines'                       | Document for Ruby instance method IO::readlines.                                 |
```

File: RI.md
Comment (peterzhu2118) [reply RI.md:137]: The "see note below" part hasn't been removed.

File: RI.md
Comment (peterzhu2118) [reply RI.md:708]: Alternatively, what do you think about removing the table and instead moving the headings for each table as headings for each command?

**Comment by BurdetteLamar** [comment]
@peterzhu2118 said:

Alternatively, what do you think about removing the table and instead moving the headings for each table as headings for each command?

Done.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Fixed.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Done.

**Comment by BurdetteLamar** [comment]
@peterzhu2118, do we need to hear from a maintainer (@drbrain or @hsbt) before merging?

**Comment by BurdetteLamar** [comment]
@peterzhu2118, will you be reviewing this?

**Comment by peterzhu2118** [comment]
I think this looks good, but I would like one of the maintainers to give input before merging.

**Comment by BurdetteLamar** [comment]
> I think this looks good, but I would like one of the maintainers to give input before merging.

Agreed. @drbrain and @hsbt, can you review or recommend a reviewer?

File: .document
Comment (colby-swandale) [inline .document:5]: 👋🏻  You identified this document's primary audience is people new to the `ri` command, and they would find this document mostly through GitHub. But, when rendering the document in GitHub, there are lots of weird and missing formatting artifacts coming through that make it challenging to read:

* Document's title header is not rendered correctly 
* Links not being rendered correctly, i.e., `{Ruby online documentation}[https://docs.ruby-lang.org/en/master]:`
* Lots of use of links using `rdoc-ref`, but these are not rendered on GitHub. Does this mean the document is intended to be read through `ri`?  How do i do that?

Perhaps it would be helpful to either commit to Markdown and it's ways of formatting, or stick with the RDoc way and continue to use `ri.rdoc`?

Comment (st0012) [reply .document:5]: > Perhaps it would be helpful to either commit to Markdown and it's ways of formatting, or stick with the RDoc way and continue to use ri.rdoc?

I'd support committing to markdown format and will be happy to help with that conversion as I did [something similar](https://github.com/ruby/irb/pull/890) for IRB just not too long ago 🙂

Comment (BurdetteLamar) [reply .document:5]: > 👋🏻 You identified this document's primary audience is people new to the `ri` command, and they would find this document mostly through GitHub. But, when rendering the document in GitHub, there are lots of weird and missing formatting artifacts coming through that make it challenging to read:
> 
>     * Document's title header is not rendered correctly
> 
>     * Links not being rendered correctly, i.e., `{Ruby online documentation}[https://docs.ruby-lang.org/en/master]:`
> 
>     * Lots of use of links using `rdoc-ref`, but these are not rendered on GitHub. Does this mean the document is intended to be read through `ri`?  How do i do that?
> 
> 
> Perhaps it would be helpful to either commit to Markdown and it's ways of formatting, or stick with the RDoc way and continue to use `ri.rdoc`?

@colby-swandale, is there a way I can view this new file as it's rendered on GitHub?  I'd like to see the problems there firsthand.

Also, not sure what you mean by 'commit to markdown'; the new page is `.md` -- a markdown page.  Maybe some of the text has an `rdoc` accent of which I'm not aware?

**Comment by BurdetteLamar** [comment]
@drbrain and @hsbt:  Are you at all interested in this work?  If so, will you (or someone you suggest) be reviewing?  If not, I think no one should spend more time on this.

File: .document
Comment (colby-swandale) [reply .document:5]: > @colby-swandale, is there a way I can view this new file as it's rendered on GitHub? I'd like to see the problems there firsthand.

Sure! The current version is at: https://github.com/ruby/rdoc/blob/21505b91dd770e31c485205ac74ed44ef832432a/RI.md

> Also, not sure what you mean by 'commit to markdown';

The page is Markdown, but the document is utilising RDoc ways of doing things, particularly around links. When Github renders the document, the links are being rendered as raw text:

Screenshot 2024-05-02 at 7 50 40 AM

---

Links in Markdown need to be denoted with `[<text>](<link>)` instead of `{<text>}[<link>]` as per the [Markdown Spec](https://www.markdownguide.org/basic-syntax/#links)

**Comment by BurdetteLamar** [comment]
Thanks much, @colby-swandale!  I'll work with the links.

(I have done some other `.md` pages that will also need attention.)

**Comment by BurdetteLamar** [comment]
Marking as draft while I work on the links (etc.).

**Comment by BurdetteLamar** [comment]
I have converted the links to markdown-style reference links:

- Those that have URLs work fine, both in the HTML and on the GitHub page.
- Those that have `rdoc-ref` work in the HTML, but not on the GitHub page.

I've experimented, and have found various forms that work on one or the other, but not on both.  
Any hints, anyone?

**Comment by BurdetteLamar** [comment]
> I have converted the links to markdown-style reference links:
> 
>     * Those that have URLs work fine, both in the HTML and on the GitHub page.
> 
>     * Those that have `rdoc-ref` work in the HTML, but not on the GitHub page.
> 
> 
> I've experimented, and have found various forms that work on one or the other, but not on both. Any hints, anyone?

@colby-swandalem, any ideas about this?

**Comment by colby-swandale** [comment]
> I have converted the links to markdown-style reference links:
> 
> * Those that have URLs work fine, both in the HTML and on the GitHub page.
> * Those that have `rdoc-ref` work in the HTML, but not on the GitHub page.
> 
> I've experimented, and have found various forms that work on one or the other, but not on both. Any hints, anyone?

I'm a bit lost sorry, if having URLs works for both renders of the document, then that's a good thing right? What issues are there going this route?

**Comment by BurdetteLamar** [comment]
> > I have converted the links to markdown-style reference links:
> > 
> > * Those that have URLs work fine, both in the HTML and on the GitHub page.
> > * Those that have `rdoc-ref` work in the HTML, but not on the GitHub page.
> > 
> > I've experimented, and have found various forms that work on one or the other, but not on both. Any hints, anyone?
> 
> I'm a bit lost sorry, if having URLs works for both renders of the document, then that's a good thing right? What issues are there going this route?

The route of all-URL, no rdoc-ref?  Would work for both, but:

- At a minimum would need to be added to the [Documentation Guide](https://github.com/ruby/ruby/blob/master/doc/contributing/documentation_guide.md);  thoughts from the Guide guide @jeremyevans?

- At a maximum, would need to be put up as an issue for the dev meeting.

Alternatively, RDoc could in its RI output convert rdoc-refs to URLs?  Out-of-scope for me, but possibly interesting.

**Comment by jeremyevans** [comment]
In general, `rdoc-ref` is useful because it makes the references internally consistent.  If all references are to URLs, then there is no need for `rdoc-ref`, but if you are referencing something else in the documentation currently being generated, you want to use `rdoc-ref` as opposed to URLs.  It's a shame GitHub doesn't handle `rdoc-ref`, but that's a GitHub issue, and I don't think we should avoid `rdoc-ref` just because GitHub doesn't render it correctly.  The goal should be something that works best with rdoc/ri generated documentation.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Fixed.

**Comment by BurdetteLamar** [comment]
Maintainers @drbrain and @hsbt, this won't be merged without your blessing, or one from someone you designate.  Can we proceed here?

File: RI.md
Comment (st0012) [inline RI.md:53]: In the latest IRB we've repurposed `help` to show the help message of IRB itself 🙂 

```suggestion
  just type `'show_doc'`.
```

File: RI.md
Comment (st0012) [inline RI.md:168]: This look similar to to https://github.com/ruby/rdoc/pull/1100/files#r1618865177 

(by @colby-swandale)

> Is the use of `\` necessary here as it's hard to tell what it's for? Should we perhaps be wrapping these in code blocks?

File: RI.md
Comment (st0012) [inline RI.md:180]:
```suggestion
  type `'show_doc'` to enter `ri`, newline to exit.
```

File: RI.md
Comment (st0012) [inline RI.md:831]: IRB now has its own dedicated doc site, let's use it:

```suggestion
[9]: https://ruby.github.io/irb/
```

File: RI.md
Comment (st0012) [inline RI.md:601]: This title and many `Options` in subtitles under this section all have backslashes, which don't seem to be necessary. Can we remove them?

**Comment by BurdetteLamar** [comment]
Maintainers @hsbt and @nobu: any comments on this PR?

**Comment by colby-swandale** [comment]
@BurdetteLamar 👋🏻  myself and @st0012 now have commit access to RDoc with the aim of becoming maintainers. I think we will be happy to merge this once the last of the feedback has been addressed.

Let me know if you have any questions!

**Comment by BurdetteLamar** [comment]
> @BurdetteLamar 👋🏻 myself and @st0012 now have commit access to RDoc with the aim of becoming maintainers. I think we will be happy to merge this once the last of the feedback has been addressed.
> 
> Let me know if you have any questions!

Thanks, @colby-swandale.  I'll try to sort out what's unaddressed.  (Won't be easy, with conversation of 77 parts!)

File: RI.md
Comment (BurdetteLamar) [reply RI.md:122]: This idiom is what I found that keeps the double-hyphen from converting to a dash.  Also, if in code block, the token would not link, which I think matters.

File: RI.md
Comment (BurdetteLamar) [reply RI.md:168]: Same as there.  Stet, I think.

File: RI.md
Comment (BurdetteLamar) [reply RI.md:831]: I disagree.  I'm writing for Ruby, and prefer to stay on its site when practicable.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Fixed? (Intermediate space added.)

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Replaced.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Fixed.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Fixed.

File: RI.md
Comment (BurdetteLamar) [reply RI.md]: Fixed.

**Comment by BurdetteLamar** [comment]
@colby-swandale and @st0012:  anything more needed here?

**Comment by BurdetteLamar** [comment]
Can I get a review, an approval, a merge here?  (At this point, I'd almost settle for a close.)

**Comment by colby-swandale** [comment]
I apologise for the wait @BurdetteLamar, but I hope you can understand that we are busy with our lives and can only volunteer so much of our time.

**Review by colby-swandale — APPROVED** [review]

File: RI.md
Comment (colby-swandale) [reply RI.md:122]: I think we will see how the community reacts to this. If it's clear what the intention is, than great. Otherwise we will probably need to review the use of these.

**Pull request merged**
